ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/ContractsReader.scala

```diff
@@ -0,0 +1,111 @@
+
+  override def lookupActiveContract(
+      submitter: Party,
+      contractId: ContractId,
+  ): Future[Option[Contract]] =
+    dispatcher.executeSql("lookup_active_contract") { implicit connection =>
+      SQL"select participant_contracts.contract_id, template_id, create_argument from #$contractsTable where contract_witness = $submitter and participant_contracts.contract_id = $contractId"
+        .as(contractRowParser.singleOpt)
+    }
+  override def lookupContractKey(
+      submitter: Party,
+      key: events.Key,
+  ): Future[Option[events.ContractId]] =
+    dispatcher.executeSql("lookup_contract_by_key") { implicit connection =>
+      lookupContractKeyQuery(submitter, key).as(contractId("contract_id").singleOpt)
+    }
+
+  override def lookupMaximumLedgerTime(ids: Set[ContractId]): Future[Option[Instant]] =
+    if (ids.isEmpty) {
+      Future.failed(emptyContractIds)
+    } else {
+      dispatcher.executeSql("lookup_maximum_ledger_time") { implicit connection =>
+        SQL"select max(create_ledger_effective_time) as max_create_ledger_effective_time, count(*) as num_contracts from participant_contracts where participant_contracts.contract_id in ($ids)"
+          .as((instant("max_create_ledger_effective_time").? ~ int("num_contracts")).single.map {
+            case result ~ retrievedContracts if retrievedContracts == ids.size => result
+            case _ => throw notFound(ids)
+          })
+      }
+    }
+
+}
+
+object ContractsReader {
+
+  private[dao] def apply(dispatcher: DbDispatcher, dbType: DbType): ContractsReader =
+    dbType match {
+      case DbType.Postgres => new Postgresql(dispatcher)
+      case DbType.H2Database => new H2Database(dispatcher)
+    }
+
+  private final class Postgresql(dispatcher: DbDispatcher) extends ContractsReader(dispatcher) {
+    override protected def lookupContractKeyQuery(
+        submitter: Party,
+        key: events.Key,
+    ): SimpleSql[Row] =
+      SQL"select participant_contracts.contract_id from #$contractsTable where $submitter =ANY(create_stakeholders) and contract_witness = $submitter and create_key_hash = ${hash(key)}"
+  }
+
+  private final class H2Database(dispatcher: DbDispatcher) extends ContractsReader(dispatcher) {
+    override protected def lookupContractKeyQuery(
+        submitter: Party,
+        key: events.Key,
+    ): SimpleSql[Row] =
+      SQL"select participant_contracts.contract_id from #$contractsTable where array_contains(create_stakeholders, $submitter) and contract_witness = $submitter and create_key_hash = ${hash(key)}"
+  }
+
+  // The contracts table _does not_ store agreement texts as they are
+  // unnecessary for interpretation and validation. The contracts returned
+  // from this table will _always_ have an empty agreement text.
+  private val contractRowParser: RowParser[Contract] =
+    str("contract_id") ~ str("template_id") ~ binaryStream("create_argument") map {
```

Comment:
I hope the `binaryStream` won't bite us with H2 again.

Reply:
Rather than hoping, I'll convert this to a `byteArray`, thanks for raising this.